File: src/dnfw/codec/aplibpack.py
"""aPLib-variant packer: produces a stream `codec.aplib.depack` accepts.

The bit-level encoding -- literal, match, offset reuse, gamma codes, the far
threshold, the end-of-stream token -- is ported exactly from `ap_pack` in
mischa85/elektron-firmware-tool (`compress.c`), MIT, with thanks. So is the
cost model (`_gamma_cost`, `match_cost`), which is what makes the parse
decisions here cost-based rather than guesswork.

**The parse strategy is deliberately not the same.** `compress.c` runs a
cost-optimal forward dynamic program over every position with a 2,048-deep
match chain. That is the right answer in C and unreachable in Python: the
Digitone II MAIN OS section is 3,085,696 bytes. This module uses greedy
matching with a one-position lazy lookahead, decided by the same cost function.

It costs nothing measurable. Against the sections Elektron ship (2026-09-11,
with the window below in force):

    DN2 1.10E bootstrap    30,302 B raw ->    16,131 vs    16,160 stock   -0.2%
    DN2 1.10E MAIN OS   3,085,696 B raw -> 1,091,713 vs 1,093,805 stock   -0.2%
    DN2 1.10E blob        833,060 B raw ->   596,466 vs   600,147 stock   -0.6%
    DN2 1.11  MAIN OS   3,192,192 B raw -> 1,129,103 vs 1,130,517 stock   -0.1%
    DN2 1.11  section 8   159,948 B raw ->   102,552 vs   103,407 stock   -0.8%

Every section comes out smaller than stock, so a rebuild never needs more flash
than the image it replaces. MAIN OS packs in under a minute.

**No match reaches back further than `limits.PACK_MAX_OFFSET`.** Without that
bound this packer reached three megabytes back, which Elektron's never do, and
the images it made stalled in recovery. The bound also made it *smaller* --
1.11 MAIN OS came out 445 bytes over stock without it -- because a long, far
match costs more bits than it saves, and faster, because the chain walk stops
at the window. See `codec.limits`.
"""

from .aplib import FAR_THRESHOLD, MIN_MATCH, OFFSET_BIAS, REUSE_GAMMA
from .limits import MAX_MATCH, PACK_MAX_OFFSET
LITERAL_BITS = 9  # 1 control bit + 8 data bits
DEFAULT_CHAIN_DEPTH = 32

# Two, not three or four. The minimum match length in this format is two, so a
# wider hash key silently discards every short match: measured on the Digitone
# II bootstrap section, a 4-byte key costs 12.4% and a 3-byte key 3.2% against stock,
# while a 2-byte key at depth 32 comes out 0.2% *smaller* than Elektron ship.
_HASH_BYTES = 2


class _Writer:
    """Interlaced bit/byte output. Control bits accumulate into a tag byte that
    is reserved in the stream at the moment its first bit is written, so bits
    and inline bytes stay in the order the depacker expects."""

    __slots__ = ("out", "tag_pos", "tag_bits")

    def __init__(self) -> None:
        self.out = bytearray()
        self.tag_pos = -1
        self.tag_bits = 0

    def bit(self, value: int) -> None:
        if self.tag_bits == 0:
            self.tag_pos = len(self.out)
            self.out.append(0)
            self.tag_bits = 8
        if value:
            self.out[self.tag_pos] |= 1 << (self.tag_bits - 1)
        self.tag_bits -= 1

    def byte(self, value: int) -> None:
        self.out.append(value)

    def gamma(self, value: int) -> None:
        width = value.bit_length()
        for i in range(width - 2, -1, -1):
            self.bit((value >> i) & 1)
            self.bit(1 if i == 0 else 0)

    def literal(self, value: int) -> None:
        self.bit(1)
        self.byte(value)

    def match(self, offset: int, length: int, last_offset: int) -> int:
        self.bit(0)
        if offset == last_offset:
            self.gamma(REUSE_GAMMA)
        else:
            raw = offset + OFFSET_BIAS
            self.gamma(raw >> 8)
            self.byte(raw & 0xFF)
            last_offset = offset
        base = length - 1 - (1 if offset > FAR_THRESHOLD else 0)
        if base <= 3:
            self.bit(base >> 1)
            self.bit(base & 1)
        else:
            self.bit(0)
            self.bit(0)
            self.gamma(base - 2)
        return last_offset

    def end(self) -> None:
        """End of stream: a match whose raw offset field decodes to the bias."""
        self.bit(0)
        self.gamma(0x1000002)
        self.byte(0xFF)


def _gamma_cost(value: int) -> int:
    return 2 * (value.bit_length() - 1)


def match_cost(offset: int, length: int, last_offset: int) -> int:
    """Cost in bits of encoding this match, by the compress.c model."""
    bits = 1 + (2 if offset == last_offset else _gamma_cost((offset + OFFSET_BIAS) >> 8) + 8)
    base = length - 1 - (1 if offset > FAR_THRESHOLD else 0)
    bits += 2 if base <= 3 else 2 + _gamma_cost(base - 2)
    return bits


def _min_match(offset: int) -> int:
    return MIN_MATCH + 1 if offset > FAR_THRESHOLD else MIN_MATCH


def _run(data: bytes, a: int, b: int, cap: int) -> int:
    """Common-prefix length of data[a:] and data[b:], capped at `cap`.

    Compared in 32-byte slices first -- a slice comparison runs at C speed,
    which is what makes this usable on a multi-megabyte image -- then refined
    byte by byte. Overlapping matches are fine: the depacker copy reads what it
    has just written, and equality on the original array is exactly the
    condition that makes that reproduce the original.
    """
    n = 0
    while n + 32 <= cap and data[a + n : a + n + 32] == data[b + n : b + n + 32]:
        n += 32
    while n < cap and data[a + n] == data[b + n]:
        n += 1
    return n


def _best(
    data: bytes,
    i: int,
    cap: int,
    head: dict,
    prev: list,
    last_offset: int,
    chain_depth: int,
    max_offset: int,
):
    """Best (offset, length, cost) at position `i`, or (0, 0, 0) for none.

    Longest wins, ties broken by cost -- so among equally long matches the one
    that reuses the last offset, or sits nearer, is preferred. No match reaches
    back further than `max_offset`; see `codec.limits` for why.
    """
    best_offset = best_length = best_cost = 0
    best_key = None

    def consider(offset: int, length: int) -> None:
        nonlocal best_offset, best_length, best_cost, best_key
        if length < _min_match(offset):
            return
        cost = match_cost(offset, length, last_offset)
        key = (-length, cost)
        if best_key is None or key < best_key:
            best_offset, best_length, best_cost, best_key = offset, length, cost, key

    if 0 < last_offset <= min(i, max_offset):
        consider(last_offset, _run(data, i - last_offset, i, cap))

    if i + _HASH_BYTES <= len(data):
        candidate = head.get(data[i : i + _HASH_BYTES], -1)
        depth = chain_depth
        while candidate >= 0 and depth:
            if i - candidate > max_offset:
                break  # the chain runs nearest-first: everything after is further
            consider(i - candidate, _run(data, candidate, i, cap))
            if best_length >= cap:
                break
            candidate = prev[candidate]
            depth -= 1

    return best_offset, best_length, best_cost


def pack(
    data: bytes,
    chain_depth: int = DEFAULT_CHAIN_DEPTH,
    max_offset: int = PACK_MAX_OFFSET,
) -> bytes:
    """Compress `data` into an aPLib stream (no section header).

    `chain_depth` trades build time for size: how many earlier positions with
    the same 2-byte prefix are considered at each step. `max_offset` is the
    furthest back any match may reach; the default keeps inside the window
    Elektron's own streams use, and lowering it is for tests.
    """
    writer = _Writer()
    size = len(data)
    if size == 0:
        writer.end()
        return bytes(writer.out)

    head: dict = {}
    prev = [-1] * size
    last_offset = 1
    i = 0

    while i < size:
        cap = min(size - i, MAX_MATCH)
        offset, length, cost = _best(
            data, i, cap, head, prev, last_offset, chain_depth, max_offset
        )

        if length and i + 1 < size:
            # Lazy: is a literal here plus the next position match cheaper per
            # byte than taking this match now?
            next_cap = min(size - i - 1, MAX_MATCH)
            _, alt_length, alt_cost = _best(
                data, i + 1, next_cap, head, prev, last_offset, chain_depth, max_offset
            )
            if alt_length and (LITERAL_BITS + alt_cost) * length < cost * (1 + alt_length):
                length = 0

        if length:
            last_offset = writer.match(offset, length, last_offset)
            step = length
        else:
            writer.literal(data[i])
            step = 1

        limit = min(i + step, size - _HASH_BYTES + 1)
        for k in range(i, limit):
            key = data[k : k + _HASH_BYTES]
            prev[k] = head.get(key, -1)
            head[key] = k
        i += step

    writer.end()
    return bytes(writer.out)

File: site/js/aplibpack.js
/**
 * aPLib-variant packer: produces a stream `aplib.depack` accepts.
 *
 * Port of `src/dnfw/codec/aplibpack.py`, whose bit-level encoding and cost
 * model come from `ap_pack` in mischa85/elektron-firmware-tool (`compress.c`),
 * MIT, with thanks. **The parse strategy is ported exactly too**, including the
 * order matches are considered in and the point at which the hash chain is
 * updated, because the goal here is not merely a valid stream but the *same*
 * stream the Python produces — that equality is what `test/test_js_codec.py`
 * asserts, and it is a far sharper check than "it round-trips".
 *
 * ## Why this exists when `packStore` already worked
 *
 * `packStore` below emits literals only. It is correct, trivially so, and it
 * was the right first move. But it costs 9/8 of the input, which took a
 * modified section 7 from 602 KB to 942 KB and the `.syx` from 2,389,280 to
 * 2,819,488 bytes — **340 KB larger than any image Elektron ship, and larger
 * than anything that has ever been flashed to a Digitone II.**
 *
 * The owner asked the obvious question: the transient entries are fixed-size
 * slots replaced in place, so why does anything grow? It doesn't. The
 * *unpacked* section is 836,956 bytes whatever you put in it. Only the
 * compressed size moves, and only because compression finds redundancy and how
 * much is there depends on the audio.
 *
 * Measured on 1.11, section 7, through this parse:
 *
 *     stock                                602,076   syx 2,389,280
 *     markers (the image flashed 09-14)    564,196   syx 2,341,280   -48,000
 *     34 realistic drum hits               576,364   syx 2,356,640   -32,640
 *     34 slots of white noise (worst)      653,852   syx 2,454,816   +65,536
 *     store-only                           941,583   syx 2,819,488  +430,208
 *
 * So the real packer stays within ±64 KB and is *smaller* for anything
 * musical — and the image confirmed on hardware was 48,000 bytes **smaller**
 * than stock. Store-only was buying a day of work and paying for it with an
 * untested hardware condition on the one path a user would actually take.
 *
 * `packStore` is kept: it is the fallback whose correctness can be argued in a
 * sentence, and it is what `pack` is checked against when a stream has to be
 * known-good rather than small.
 */

import { Growable } from "./bytes.js";
import { FAR_THRESHOLD, MIN_MATCH, OFFSET_BIAS, REUSE_GAMMA } from "./aplib.js";
import { MAX_MATCH, PACK_MAX_OFFSET } from "./limits.js";

const LITERAL_BITS = 9;            // 1 control bit + 8 data bits
export const DEFAULT_CHAIN_DEPTH = 32;

/**
 * Two, not three or four. The minimum match in this format is two, so a wider
 * hash key silently discards every short match: measured on the Digitone II
 * bootstrap section, a 4-byte key costs 12.4% and a 3-byte key 3.2% against
 * stock, while a 2-byte key at depth 32 comes out 0.2% *smaller* than Elektron
 * ship.
 */
const HASH_BYTES = 2;
const HASH_SLOTS = 1 << 16;        // a 2-byte key indexes this directly

/**
 * Interlaced bit/byte output. Control bits accumulate into a tag byte reserved
 * in the stream the moment its first bit is written, so bits and inline bytes
 * stay in the order the depacker expects.
 */
class Writer {
  constructor(capacity) {
    this.out = new Growable(capacity);
    this.tagPos = -1;
    this.tagBits = 0;
  }

  bit(value) {
    if (this.tagBits === 0) {
      this.tagPos = this.out.length;
      this.out.push(0);
      this.tagBits = 8;
    }
    if (value) this.out.buf[this.tagPos] |= 1 << (this.tagBits - 1);
    this.tagBits -= 1;
  }

  /**
   * Elias gamma, interlaced. Written without assuming the value fits a 32-bit
   * shift: the end-of-stream token's gamma is 0x1000002.
   */
  gamma(value) {
    let width = 0;
    for (let v = value; v > 1; v = Math.floor(v / 2)) width += 1;
    for (let i = width - 1; i >= 0; i--) {
      this.bit(Math.floor(value / 2 ** i) % 2);
      this.bit(i === 0 ? 1 : 0);
    }
  }

  literal(value) {
    this.bit(1);
    this.out.push(value);
  }

  /** -> the new last-offset, which a match may reuse instead of encoding. */
  match(offset, length, lastOffset) {
    this.bit(0);
    if (offset === lastOffset) {
      this.gamma(REUSE_GAMMA);
    } else {
      const raw = offset + OFFSET_BIAS;
      this.gamma(raw >>> 8);
      this.out.push(raw & 0xff);
      lastOffset = offset;
    }
    const base = length - 1 - (offset > FAR_THRESHOLD ? 1 : 0);
    if (base <= 3) {
      this.bit(base >> 1);
      this.bit(base & 1);
    } else {
      this.bit(0);
      this.bit(0);
      this.gamma(base - 2);
    }
    return lastOffset;
  }

  /** End of stream: a match whose raw offset field decodes to the bias. */
  end() {
    this.bit(0);
    this.gamma(0x1000002);
    this.out.push(0xff);
  }
}

const bitLength = (value) => 32 - Math.clz32(value);
const gammaCost = (value) => 2 * (bitLength(value) - 1);

/** Cost in bits of encoding this match, by the compress.c model. */
export function matchCost(offset, length, lastOffset) {
  let bits = 1 + (offset === lastOffset
    ? 2
    : gammaCost((offset + OFFSET_BIAS) >>> 8) + 8);
  const base = length - 1 - (offset > FAR_THRESHOLD ? 1 : 0);
  bits += base <= 3 ? 2 : 2 + gammaCost(base - 2);
  return bits;
}

const minMatch = (offset) => (offset > FAR_THRESHOLD ? MIN_MATCH + 1 : MIN_MATCH);

/**
 * Common-prefix length of data[a:] and data[b:], capped at `cap`.
 *
 * Overlapping matches are fine: the depacker's copy reads what it has just
 * written, and equality on the original array is exactly the condition that
 * makes that reproduce the original.
 */
function run(data, a, b, cap) {
  let n = 0;
  while (n < cap && data[a + n] === data[b + n]) n += 1;
  return n;
}

/**
 * Best (offset, length, cost) at position `i`, or zeros for none.
 *
 * Longest wins, ties broken by cost — so among equally long matches the one
 * that reuses the last offset, or sits nearer, is preferred. No match reaches
 * back further than `maxOffset`; see `limits.js`.
 *
 * Results go into the caller's `found` array rather than a fresh object: this
 * runs once or twice per input byte over megabytes, and allocating there is the
 * difference between a browser tab that works and one that stutters.
 */
function best(data, i, cap, head, prev, lastOffset, chainDepth, maxOffset, found) {
  let bestOffset = 0;
  let bestLength = 0;
  let bestCost = 0;
  let haveBest = false;

  const consider = (offset, length) => {
    if (length < minMatch(offset)) return;
    const cost = matchCost(offset, length, lastOffset);
    // The Python compares the tuple (-length, cost): longer first, then cheaper.
    if (!haveBest || length > bestLength || (length === bestLength && cost < bestCost)) {
      bestOffset = offset;
      bestLength = length;
      bestCost = cost;
      haveBest = true;
    }
  };

  if (lastOffset > 0 && lastOffset <= Math.min(i, maxOffset)) {
    consider(lastOffset, run(data, i - lastOffset, i, cap));
  }

  if (i + HASH_BYTES <= data.length) {
    let candidate = head[(data[i] << 8) | data[i + 1]];
    let depth = chainDepth;
    while (candidate >= 0 && depth) {
      // The chain runs nearest-first, so everything after this is further too.
      if (i - candidate > maxOffset) break;
      consider(i - candidate, run(data, candidate, i, cap));
      if (bestLength >= cap) break;
      candidate = prev[candidate];
      depth -= 1;
    }
  }

  found[0] = bestOffset;
  found[1] = bestLength;
  found[2] = bestCost;
}

/**
 * Compress `data` into an aPLib stream (no section header) -> Uint8Array.
 *
 * `chainDepth` trades time for size: how many earlier positions with the same
 * 2-byte prefix are considered at each step. `maxOffset` is the furthest back
 * any match may reach; the default keeps inside the window Elektron's own
 * streams use, and lowering it is for tests.
 */
export function pack(data, { chainDepth = DEFAULT_CHAIN_DEPTH,
                             maxOffset = PACK_MAX_OFFSET } = {}) {
  const size = data.length;
  // 9/8 plus slack: a literal-only parse is the worst this can do, so the
  // buffer never has to grow on real input.
  const writer = new Writer(Math.ceil(size * 1.15) + 64);
  if (size === 0) {
    writer.end();
    return writer.out.bytes();
  }

  // Int32Array rather than a Map: a 2-byte key indexes 65,536 slots directly,
  // and `prev` is one entry per input position. -1 means "no earlier match".
  const head = new Int32Array(HASH_SLOTS).fill(-1);
  const prev = new Int32Array(size).fill(-1);

  const here = new Int32Array(3);
  const ahead = new Int32Array(3);
  let lastOffset = 1;
  let i = 0;

  while (i < size) {
    const cap = Math.min(size - i, MAX_MATCH);
    best(data, i, cap, head, prev, lastOffset, chainDepth, maxOffset, here);
    let length = here[1];

    if (length && i + 1 < size) {
      // Lazy: is a literal here plus the next position's match cheaper per byte
      // than taking this match now? Note the hash chain has NOT yet been
      // updated for position i, matching the Python exactly -- the lookahead
      // cannot see a match starting where we are standing.
      const nextCap = Math.min(size - i - 1, MAX_MATCH);
      best(data, i + 1, nextCap, head, prev, lastOffset, chainDepth, maxOffset, ahead);
      if (ahead[1] && (LITERAL_BITS + ahead[2]) * length < here[2] * (1 + ahead[1])) {
        length = 0;
      }
    }

    let step;
    if (length) {
      lastOffset = writer.match(here[0], length, lastOffset);
      step = length;
    } else {
      writer.literal(data[i]);
      step = 1;
    }

    const limit = Math.min(i + step, size - HASH_BYTES + 1);
    for (let k = i; k < limit; k++) {
      const key = (data[k] << 8) | data[k + 1];
      prev[k] = head[key];
      head[key] = k;
    }
    i += step;
  }

  writer.end();
  return writer.out.bytes();
}

/**
 * Literals only: one control bit per byte, plus the byte. 9/8 of the input.
 *
 * Kept as the fallback whose correctness can be argued in a sentence — it
 * exercises one token type and the terminator, with no offsets, no lengths and
 * no reuse state. `pack` is what ships; this is what `pack` is measured
 * against when a stream has to be known-good rather than small.
 */
export function packStore(data) {
  const writer = new Writer(Math.ceil(data.length * 1.15) + 64);
  for (let i = 0; i < data.length; i++) writer.literal(data[i]);
  writer.end();
  return writer.out.bytes();
}
